File: src/main/java/com/hileco/cortex/constraints/expressions/Expression.java
package com.hileco.cortex.constraints.expressions;

import com.hileco.cortex.vm.ProgramStoreZone;
import com.microsoft.z3.BoolExpr;
import com.microsoft.z3.Context;
import com.microsoft.z3.Expr;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public interface Expression {
    Expr asZ3Expr(Context context, ReferenceMapping referenceMapping);

    @FunctionalInterface
    interface InputConverter {
        Expr convert(Context context, Expr input);
    }

    @FunctionalInterface
    interface LeftRightConverter {
        Expr convert(Context context, Expr left, Expr right);
    }

    @Data
    @AllArgsConstructor
    class Input implements Expression {
        private String representation;
        private InputConverter converter;
        private Expression input;

        public static Function<ExpressionStack, Expression> builder(String representation, InputConverter converter) {
            return expressionStack -> {
                var input = expressionStack.pop(0);
                return new Input(representation, converter, input);
            };
        }

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            var inputExpr = this.input.asZ3Expr(context, referenceMapping);
            return this.converter.convert(context, inputExpr);
        }

        @Override
        public String toString() {
            return String.format("%s(%s)", this.representation, this.input);
        }
    }

    @Data
    @AllArgsConstructor
    class LeftRight implements Expression {
        private String representation;
        private LeftRightConverter converter;
        private Expression left;
        private Expression right;

        public static Function<ExpressionStack, Expression> builder(String representation, LeftRightConverter converter) {
            return expressionStack -> {
                var left = expressionStack.pop(0);
                var right = expressionStack.pop(0);
                return new LeftRight(representation, converter, left, right);
            };
        }

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            var leftExpr = this.left.asZ3Expr(context, referenceMapping);
            var rightExpr = this.right.asZ3Expr(context, referenceMapping);
            return this.converter.convert(context, leftExpr, rightExpr);
        }

        @Override
        public String toString() {
            return String.format("(%s %s %s)", this.left, this.representation, this.right);
        }
    }

    @Data
    @AllArgsConstructor
    class Reference implements Expression {
        private ProgramStoreZone type;
        private Expression address;

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            var reference = referenceMapping.getReferencesForward().computeIfAbsent(Reference.this, unmappedReference -> {
                var key = Integer.toString(referenceMapping.getReferencesForward().size());
                referenceMapping.getReferencesBackward().put(key, unmappedReference);
                return key;
            });
            var referenceSymbol = context.mkSymbol(reference);
            return context.mkIntConst(referenceSymbol);
        }

        @Override
        public String toString() {
            return String.format("%s[%s]", this.type, this.address);
        }
    }

    @Data
    @AllArgsConstructor
    class Value implements Expression {
        private Long constant;

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            return context.mkInt(this.constant);
        }

        @Override
        public String toString() {
            return Long.toString(this.constant);
        }
    }

    @Data
    @AllArgsConstructor
    class Stack implements Expression {
        private int address;

        @Override
        public String toString() {
            return String.format("STACK[%d]", this.address);
        }

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            throw new IllegalArgumentException(String.format("Missing stack: %d.", this.address));
        }
    }

    @Data
    @AllArgsConstructor
    class Not implements Expression {
        private Expression input;

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            var inputExpr = this.input.asZ3Expr(context, referenceMapping);
            return context.mkNot((BoolExpr) inputExpr);
        }

        @Override
        public String toString() {
            return String.format("!(%s)", this.input);
        }
    }

    @Data
    @AllArgsConstructor
    class And implements Expression {
        private List<Expression> inputs;

        @Override
        public Expr asZ3Expr(Context context, ReferenceMapping referenceMapping) {
            return context.mkAnd((BoolExpr[]) this.inputs.stream()
                    .map(input -> input.asZ3Expr(context, referenceMapping))
                    .toArray(BoolExpr[]::new));
        }

        @Override
        public String toString() {
            return this.inputs.stream().map(input -> String.format("(%s)", input.toString())).collect(Collectors.joining(" && "));
        }
    }
}

File: src/main/java/com/hileco/cortex/constraints/expressions/Expression.kt
package com.hileco.cortex.constraints.expressions

import com.hileco.cortex.vm.ProgramStoreZone
import com.microsoft.z3.*
import java.lang.Long.toString

interface Expression {
    fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr

    data class Reference(val type: ProgramStoreZone,
                         val address: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val reference = referenceMapping.referencesForward.computeIfAbsent(this@Reference) { unmappedReference ->
                val key = Integer.toString(referenceMapping.referencesForward.size)
                referenceMapping.referencesBackward[key] = unmappedReference
                key
            }
            val referenceSymbol = context.mkSymbol(reference)
            return context.mkIntConst(referenceSymbol)
        }

        override fun toString(): String {
            return "$type[$address]"
        }
    }

    data class Value(val constant: Long) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            return context.mkInt(constant)
        }

        override fun toString(): String {
            return toString(constant)
        }
    }

    data class Stack(val address: Int = 0) : Expression {
        override fun toString(): String {
            return "STACK[$address]"
        }

        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            throw IllegalArgumentException("Missing stack: $address.")
        }
    }

    data class Not(val input: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val inputExpr = input.asZ3Expr(context, referenceMapping)
            return context.mkNot(inputExpr as BoolExpr)
        }

        override fun toString(): String {
            return "!($input)"
        }
    }

    data class And(val inputs: List<Expression>) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            return context.mkAnd(*inputs.map { it.asZ3Expr(context, referenceMapping) as BoolExpr }.toTypedArray())
        }

        override fun toString(): String {
            return inputs.joinToString(separator = " && ", prefix = "(", postfix = ")") { "$it" }
        }
    }

    data class Or(val inputs: List<Expression>) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            return context.mkOr(*inputs.map { it.asZ3Expr(context, referenceMapping) as BoolExpr }.toTypedArray())
        }

        override fun toString(): String {
            return inputs.joinToString(separator = " || ", prefix = "(", postfix = ")") { "$it" }
        }
    }

    object True : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            return context.mkTrue()
        }

        override fun toString(): String {
            return "TRUE"
        }
    }

    object False : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            return context.mkFalse()
        }

        override fun toString(): String {
            return "FALSE"
        }
    }

    data class Add(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkAdd(leftExpr as ArithExpr, rightExpr as ArithExpr)
        }

        override fun toString(): String {
            return "($left + $right)"
        }
    }

    data class Subtract(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkSub(leftExpr as ArithExpr, rightExpr as ArithExpr)
        }

        override fun toString(): String {
            return "($left - $right)"
        }
    }

    data class Multiply(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkMul(leftExpr as ArithExpr, rightExpr as ArithExpr)
        }

        override fun toString(): String {
            return "($left * $right)"
        }
    }

    data class Divide(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkDiv(leftExpr as ArithExpr, rightExpr as ArithExpr)
        }

        override fun toString(): String {
            return "($left / $right)"
        }
    }

    data class LessThan(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkLt(leftExpr as ArithExpr, rightExpr as ArithExpr)
        }

        override fun toString(): String {
            return "($left < $right)"
        }
    }

    data class GreaterThan(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkGt(leftExpr as ArithExpr, rightExpr as ArithExpr)
        }

        override fun toString(): String {
            return "($left > $right)"
        }
    }

    data class Equals(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkEq(leftExpr, rightExpr)
        }

        override fun toString(): String {
            return "($left == $right)"
        }
    }

    data class BitwiseOr(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkOr(leftExpr as BoolExpr, rightExpr as BoolExpr)
        }

        override fun toString(): String {
            return "($left || $right)"
        }
    }

    data class BitwiseAnd(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkAnd(leftExpr as BoolExpr, rightExpr as BoolExpr)
        }

        override fun toString(): String {
            return "($left && $right)"
        }
    }

    data class Modulo(val left: Expression, val right: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val leftExpr = left.asZ3Expr(context, referenceMapping)
            val rightExpr = right.asZ3Expr(context, referenceMapping)
            return context.mkMod(leftExpr as IntExpr, rightExpr as IntExpr)
        }

        override fun toString(): String {
            return "($left % $right)"
        }
    }

    data class IsZero(val input: Expression) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val inputExpr = input.asZ3Expr(context, referenceMapping)
            return if (inputExpr.isBool) {
                context.mkEq(inputExpr, context.mkBool(false))
            } else {
                context.mkEq(inputExpr, context.mkInt(0))
            }
        }

        override fun toString(): String {
            return "0 == $input"
        }
    }

    data class Hash(val input: Expression, val method: String) : Expression {
        override fun asZ3Expr(context: Context, referenceMapping: ReferenceMapping): Expr {
            val inputExpr = input.asZ3Expr(context, referenceMapping)
            val hashFunction = referenceMapping.hashFunctions.getOrPut("HASH_$method") {
                // TODO encode that x != y implies HASH(x) != HASH(y)
                context.mkFuncDecl("HASH_$method", context.intSort, context.intSort)
            }
            return context.mkApp(hashFunction, inputExpr)
        }

        override fun toString(): String {
            return "HASH_$method($input)"
        }
    }

    companion object {
        private val IS_EQUIVALENT_TRUE: (Expression) -> Boolean = { it == True || (it is Expression.Value && it.constant >= 0) }
        private val IS_EQUIVALENT_FALSE: (Expression) -> Boolean = { it == False || it == Value(0) }
        fun constructAnd(inputs: List<Expression>): Expression {
            val distinctInputs = inputs.distinct().filterNot(IS_EQUIVALENT_TRUE)
            val falseInputs = distinctInputs.count(IS_EQUIVALENT_FALSE)
            when {
                falseInputs > 0 -> return False
                distinctInputs.isEmpty() -> return True
                distinctInputs.size == 1 -> return distinctInputs.single()
            }
            return And(distinctInputs)
        }

        fun constructOr(inputs: List<Expression>): Expression {
            val distinctInputs = inputs.distinct().filterNot(IS_EQUIVALENT_FALSE)
            val trueInputs = distinctInputs.count(IS_EQUIVALENT_TRUE)
            when {
                trueInputs > 0 -> return True
                distinctInputs.isEmpty() -> return True
                distinctInputs.size == 1 -> return distinctInputs.single()
            }
            return Or(distinctInputs)
        }
    }
}
